docs/proposals/20210323-etcdadm-bootstrap-provider.md

```diff
@@ -0,0 +1,556 @@
+    // Total number of non-terminated machines targeted by this etcd cluster
+    // (their labels match the selector).
+    // +optional
+    ReadyReplicas int32 `json:"replicas,omitempty"`
+
+    // +optional
+    InitMachineAddress string `json:"initMachineAddress"`
+
+    // +optional
+    Initialized bool `json:"initialized"`
+
+    // +optional
+    Ready bool `json:"ready"`
+
+    // +optional
+    Endpoint string `json:"endpoint"`
+
+    // +optional
+    Selector string `json:"selector,omitempty"`
+}
+
+// API Group: bootstrap.cluster.x-k8s.io
+type EtcdadmConfig struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+    
+    Spec   EtcdadmConfigSpec   `json:"spec,omitempty"`
+    Status EtcdadmConfigStatus `json:"status,omitempty"`
+}
+
+type EtcdadmConfigSpec struct {
+    // Users specifies extra users to add
+    // +optional
+    Users []capbk.User `json:"users,omitempty"`
+    // PreEtcdadmCommands specifies extra commands to run before kubeadm runs
+    // +optional
+    PreEtcdadmCommands []string `json:"preEtcdadmCommands,omitempty"`
+
+    // PostEtcdadmCommands specifies extra commands to run after kubeadm runs
+    // +optional
+    PostEtcdadmCommands []string `json:"postEtcdadmCommands,omitempty"`
+
+    // +optional
+    Version string `json:"version,omitempty"`
+
+    // +optional
+    EtcdadmArgs map[string]interface{} `json:"etcdadmArgs,omitempty"`
+}
+
+
+type EtcdadmConfigStatus struct {
+    Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+    
+    DataSecretName *string `json:"dataSecretName,omitempty"`
+
+    Ready bool `json:"ready,omitempty"`
+}
+```
+
+The following fields will be added/modified in existing CAPI types:
+- Add Cluster.Spec.ManagedExternalEtcdRef as:
+```go
+// ManagedExternalEtcdRef is an optional reference to an etcd provider resource that holds details
+// for provisioning an external etcd cluster
+// +optional
+ManagedExternalEtcdRef *corev1.ObjectReference `json:"managedExternalEtcdRef,omitempty"`
+```
+- Add Cluster.Status.ManagedExternalEtcdReady field as:
+```go
+// ManagedExternalEtcdReady indicates external etcd cluster is fully provisioned
+// +optional
+ManagedExternalEtcdReady bool `json:"managedExternalEtcdReady"`
+```
+- Make KubeadmConfigSpec.ClusterConfiguration.Etcd.External.Endpoints field mutable
+    - The kubeadm validating webhook currently doesn't allow this field to get updated. It should be added as an [allowed path](https://github.com/mrajashree/cluster-api/commit/18c42c47d024ce17cdda39500fc0d6bd67c5aa67#diff-6603d336435f3ee62a50043413f62441b685d5c416e5cadd792ed2a536b2f55fR121) in the webhook.
+
+
+#### Implementation details
+
+**Contract for pluggable etcd providers**
+
+##### An etcd controller’s main responsibilites are:
+
+* Managing a set of Machines that represent an etcd member
+* Managing the etcd cluster, in terms of scaling up/down/upgrading etcd version, and running healthchecks
```

Comment:
Can we clarify which component is doing the health check? Is that from etcd-manager to each node? Or peer to peer?